include/Basic/SourceManager.hpp

```diff
@@ -0,0 +1,273 @@
+#ifndef GLU_SOURCE_MANAGER_HPP
+#define GLU_SOURCE_MANAGER_HPP
+
+#include <llvm/ADT/SmallVector.h>
+#include <llvm/ADT/StringRef.h>
+#include <llvm/Support/MemoryBuffer.h>
+#include <llvm/Support/VirtualFileSystem.h>
+
+#include "SourceLocation.hpp"
+
+namespace glu {
+
+///
+/// @class ContentCache
+/// @brief A file loaded will be cached in this class.
+///
+/// The ContentCache class is responsible for caching the content of a file that
+/// has been loaded. It is used by the SourceManager to store the content of a
+/// file. The content of a file is stored in a MemoryBuffer object which is
+/// owned by the ContentCache.
+///
+class alignas(8) ContentCache {
+
+private:
+    llvm::StringRef _fileName;
+
+    /// The buffer containing the content of the file.
+    mutable std::unique_ptr<llvm::MemoryBuffer> _buffer;
+    /// A flag to indicate if the buffer is invalid.
+    mutable unsigned _isBufferInvalid : 1;
```

Comment:
why use a `ContentCache const *` then use `mutable` fields... just stop using const?

Reply:
and why not directly use an std::optional... or just set the buffer to nullptr when it's invalid